File: Backtracking/216. Combination Sum III/solution.ts
function combinationSum3(k: number, n: number): number[][] {
    const result: number[][] = [];

    function backtrack(start: number, combination: number[], sumSoFar: number): void {
        if (combination.length === k && sumSoFar === n) {
            result.push([...combination]);
            return;
        }

        if (combination.length === k || sumSoFar > n) {
            return;
        }

        for (let num = start; num < 10; num++) {
            combination.push(num);
            backtrack(num + 1, combination, sumSoFar + num);
            combination.pop();
        }
    }

    backtrack(1, [], 0);
    return result;
}

File: Backtracking/216. Combination Sum III/solution.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        result = []

        def backtrack(start: int, combination: List[int], sum_so_far: int):
            if len(combination) == k and sum_so_far == n:
                result.append(combination[:])
                return

            if len(combination) == k or sum_so_far > n:
                return

            for num in range(start, 10):
                combination.append(num)
                backtrack(num + 1, combination, sum_so_far + num)
                combination.pop()

        backtrack(1, [], 0)
        return result
